### ISSUE TITLE
Livelock when consuming from wrapped body stream

### ISSUE THREAD

**Issue by mfikes** [ztellman/aleph#231]
I had previously mentioned a livelock issue in https://github.com/ztellman/aleph/issues/230#issuecomment-201917017

I'm unable to produce a minimal repro that doesn't involve any downstream code, but I'll share the details that I have found, in hopes that it is useful.

I have code (which is using Bidi/Yada with Aleph, with `raw-stream?` set to `true`), which fairly reliably exhibits a livelock (where data being sent via curl stops being sent, CPU usage drops to zero, _etc._). This is all on OS X.

This occurs with the current Aleph betas but does not occur with `0.4.0`.

When the livelock occurs (which is usually within minutes of starting a large HTTP multipart POST, but can also occur with small POSTs, if I try enough of them), it always involves the thread reading from the `InputStream` being stuck with this stack trace (with the bottom of the stack elided—it involves Cheshire consuming from the steam):

```
Name: Thread-42610
State: WAITING on java.util.concurrent.CountDownLatch$Sync@1341473
Total blocked: 0  Total waited: 17

Stack trace: 
sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
manifold.deferred.Deferred.deref(deferred.clj:439)
clojure.core$deref.invokeStatic(core.clj:2228)
clojure.core$deref.invoke(core.clj:2214)
byte_streams.pushback_stream.SynchronizedPushbackByteStream.read(pushback_stream.clj:113)
byte_streams.InputStream.read(InputStream.java:50)
sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
   - locked java.io.InputStreamReader@14903a84
java.io.InputStreamReader.read(InputStreamReader.java:184)
java.io.BufferedReader.fill(BufferedReader.java:161)
java.io.BufferedReader.read1(BufferedReader.java:212)
java.io.BufferedReader.read(BufferedReader.java:286)
   - locked java.io.InputStreamReader@14903a84
```

My code essentially takes the `body-stream` and converts it to an `InputStream` 

```
(bs/to-input-stream body-stream) 
```

This `InputStream` is then consumed in my code, (it is a stream of JSON, parsed with Cheshire's `parsed-seq`). This is done in a separate thread, with my code returning a `promise` immediately to Yada's handler with the `promise` being delivered by the parsing thread when the stream has been consumed.

What is crucial to this is that `body-stream` is not `(:body req)`, but involves a little extra wrapping being done in Yada here (including @malcolmsparks as he may be interested in this):

https://github.com/juxt/yada/blob/5e2b160027d33979571b5211363a26da0b6bbe43/src/yada/interceptors.clj#L182

So, my code ends up consuming from an input stream that effectively is built using code like

```
(bs/to-input-stream (stream/map bs/to-byte-array (bs/to-byte-buffers (:body req))))
```

as opposed to the simpler

```
(bs/to-input-stream (:body req)) 
```

If I revise the Yada code to eliminate the extra conversions, my issue reliably goes away (never occurs), and with the wrappings the issue reliably occurs.

(As an aside, this is issue no longer affects me, as I'm now using a mechanism in Yada that allows me to specify a consumer function which then operates directly on `(:body req)`, but logging this issue nevertheless, because I'm fairly certain a bug is lurking in here that would benefit being squashed.)

**Comment by ztellman** [comment]
Just to confirm, this is still true with the latest beta release?

**Comment by mfikes** [comment]
Yes, with `[aleph "0.4.1-beta7"]` and `[byte-streams "0.2.1-alpha1"]`.

**Comment by mfikes** [comment]
Crap. I have the wrong `byte-streams` version. Hold on @ztellman let me re-confirm this with `0.2.2-alpha1`

**Comment by mfikes** [comment]
Yep. PEBCAK. I owe another beer.
